stdlib/public/core/AutoDiff.swift

```diff
@@ -885,3 +885,76 @@ public struct AnyDerivative : Differentiable & AdditiveArithmetic {
     _box._move(along: direction._box)
   }
 }
+
+//===----------------------------------------------------------------------===//
+// Differentiable higher order functions for collections
+//===----------------------------------------------------------------------===//
+
+public extension Array where Element: Differentiable {
+  @differentiable(wrt: (self, initialResult), vjp: _vjpDifferentiableReduce)
+  func differentiableReduce<Result: Differentiable>(
+    _ initialResult: Result,
+    _ nextPartialResult: @differentiable (Result, Element) -> Result
+  ) -> Result {
+    reduce(initialResult, nextPartialResult)
+  }
+
+  @usableFromInline
+  internal func _vjpDifferentiableReduce<Result: Differentiable>(
+    _ initialResult: Result,
+    _ nextPartialResult: @differentiable (Result, Element) -> Result
+  ) -> (value: Result,
+        pullback: (Result.TangentVector)
+          -> (Array.TangentVector, Result.TangentVector)) {
+    var pullbacks:
+      [(Result.TangentVector) -> (Result.TangentVector, Element.TangentVector)]
+        = []
+    let count = self.count
+    pullbacks.reserveCapacity(count)
+    var result = initialResult
+    for element in self {
+      let (y, pb) =
+        Swift.valueWithPullback(at: result, element, in: nextPartialResult)
```

Comment:
Interesting why is the `Swift.` necessary?

Reply:
There is a method version of `valueWithPullback(at:in:)` this defined in a protocol extension for `Differentiable`, but we are not using that.

Reply:
Ahh okay